ghcide/src/Development/IDE/Plugin/Completions/Logic.hs

```diff
@@ -370,65 +316,63 @@ cacheDataProducer uri env curMod globalEnv inScopeEnv limports = do
 
       rdrElts = globalRdrEnvElts globalEnv
 
-      foldMapM :: (Foldable f, Monad m, Monoid b) => (a -> m b) -> f a -> m b
-      foldMapM f xs = foldr step return xs mempty where
-        step x r z = f x >>= \y -> r $! z `mappend` y
+      -- construct a map from Parents(type) to their fields
+      fieldMap = Map.fromListWith (++) $ flip mapMaybe rdrElts $ \elt -> do
```

Comment:
Can this get big? Could this repeated `++` be a performance hit?

Reply:
It's a map from type constructor to field names, so `[(Foo,[field1, field2,...]), ...]`. I guess it could get big if you have big records, but I don't think its any worse than any of the other listy code we have in completions.

Reply:
Worth a comment, casual use of `++` makes me jumpy these days :D